api/api/src/main/scala/uk/ac/wellcome/platform/api/services/FilteredAggregationBuilder.scala

```diff
@@ -0,0 +1,97 @@
+package uk.ac.wellcome.platform.api.services
+
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s.requests.searches.aggs.{
+  AbstractAggregation,
+  Aggregation,
+  FilterAggregation
+}
+import com.sksamuel.elastic4s.requests.searches.queries.Query
+import uk.ac.wellcome.display.models.AggregationRequest
+import uk.ac.wellcome.platform.api.models._
+
+import scala.collection.immutable._
+
+/** This class governs the way in which we wish to combine the filters and aggregations
+  * that are specified for a search. We have a concept of "pairing" a filter and an aggregation:
+  * for example, an aggregation on workType is paired with a filter of a specific workType.
+  * If a search includes an aggregation and its paired filter, the filter is *not* applied to that
+  * aggregation, but is still applied to results and to all other aggregations.
+  *
+  * Given a list of aggregations requests and filters, as well as functions to convert these to
+  * constituents of the ES query, this class exposes:
+  *
+  * - `aggregations`: a list of all the ES query aggregations, where those that need to be filtered
+  *   now have a sub-aggregation of the filter aggregation type, named "filtered".
+  * - `independentFilters`: a list of all of the given filters which are not paired to any of
+  *   the given aggregations. These can be used as the ES query filters.
+  * - `aggregationDependentFilters`: a list of all of the given filters which are paired to one
+  *   of the given aggregations. These can be used as the ES post-query filters.
+  */
+class FilteredAggregationBuilder(
```

Comment:
Nit as I know the naming here is fiddly: Calling this `FiltersAndAggregationsBuilder` helped me understand why they're coupled in the implementing code.

Reply:
```scala
class FiltersAndAggregationsBuilder {
  val unpairedFilters: List[WorkTypeFilter]
  val filteredAggregations: Aggregations
  val pairedFilters: List[WorkTypeFilter]
}
```

Having the names reflected in all places helped too - and made the mapping of:
```scala
request
.filters { filtersAndAggregations.unpairedFilters }
.aggregations { filtersAndAggregations.unpairedFilters }
.postFilters { filtersAndAggregations.pairedFilters }
```

feel easier to understand.

Subject to a high level of subjectivitiy, take it or leave it the functionality is the same and great.